flax/core/lift.py

```diff
@@ -719,70 +719,100 @@ def scanned(broadcast_vars, carry, scan_variable_groups, rng_groups, args):
       name='scan')
 
 
-def custom_vjp(fn: Callable[..., Any], backward_fn: Callable[..., Any],
-               grad_kind: CollectionFilter = 'params',
+def custom_vjp(fn: Callable[..., Any], 
+               forward_fn: Callable[..., Any],
+               backward_fn: Callable[..., Any],
+               grad_vars: CollectionFilter = 'params',
                nondiff_argnums=()):
-  """"Lifted version of `jax.custom_vjp`.
+  """Lifted version of `jax.custom_vjp`.
 
-  `backward_fn` defines a custom vjp (backward gradient) for `fn`.
+  `forward_fn` and `backward_fn` together define a custom vjp for `fn`.
+  The original `fn` will run in case a vjp (backward gradient) is not computed.
+
+  The `forward_fn` receives the same arguments as `fn` but is expected to return
+  a tuple containing the output of `fn(scope, *args)` and the residuals that are
+  passed to `backward_fn`.
+
+  The `backward_fn` receives the nondiff arguments, residuals, and the output tangents.
+  It should return a tuple containing the input and variable tangents.
+
+  Note that the vjp function returned by `lift.vjp` can be passed as residual and
+  used in the `backward_fn`. The scope is unavailable during the backward pass.
+  If the scope is required in `backward_fn`, a snapshot of the variables can be
+  taken and returned as a residual in the `forward_fn`.
 
   Example::
 
+    f = nn.dense
+
     def fwd(scope, x, features):
-      y = nn.dense(scope, x, features)
-      return y, x
+      y, vjp_fn = lift.vjp(partial(f, features=features), scope, x)
+      return y, vjp_fn
```

Comment:
Cool, the use of a VJP function as a VJP's residual is something I added to JAX :) https://github.com/google/jax/pull/3705